beluga/include/beluga/motion/differential_drive_model.hpp

```diff
@@ -96,7 +104,28 @@ class DifferentialDriveModel {
    * \return a callable satisfying \ref StateSamplingFunctionPage.
    */
   template <class Control, typename = common_tuple_type_t<Control, control_type>>
-  [[nodiscard]] auto operator()(Control&& action) const {
+  [[nodiscard]] auto operator()(const Control& action) const {
+    if constexpr (NDim == 2) {
+      return sampling_fn_2d(action);
+    } else {
+      return sampling_fn_3d(action);
+    }
+  }
+
+ private:
+  using control_type_2d = std::tuple<Sophus::SE2d, Sophus::SE2d>;
+  using control_type_3d = std::tuple<Sophus::SE3d, Sophus::SE3d>;
+
+  template <class Control, typename = common_tuple_type_t<Control, control_type_3d>>
+  [[nodiscard]] auto sampling_fn_3d(const Control& action) const {
+    const auto& [current_odom, previous_odom] = action;
```

Comment:
nit: Using this for the internals might save a little in terms of template instantiations.
```suggestion
  [[nodiscard]] auto sampling_fn_3d(const Sophus::SE3d& current_odom, const Sophus::SE3d& previous_odom) const {
```